src/cli/mod.rs

```diff
@@ -0,0 +1,49 @@
+// Copyright 2020 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Base CLI implementation.
+
+use crate::common::{PROJECT_AUTHOR, PROJECT_DESC, PROJECT_NAME, PROJECT_VERSION};
+use crate::error::ParsecToolError;
+use crate::subcommands::{ParsecToolSubcommand, Subcommand};
+use parsec_client::auth::AuthenticationData;
+use structopt::StructOpt;
+
+/// Struct representing the command-line interface of parsec-tool.
+#[derive(Debug, StructOpt)]
+#[structopt(name=PROJECT_NAME, about=PROJECT_DESC, author=PROJECT_AUTHOR, version=PROJECT_VERSION)]
+pub struct ParsecToolApp {
+    /// How verbose should we be?
+    #[structopt(short = "v", multiple = true, default_value = "0")]
+    pub verbosity: u8,
+
+    /// Sets the authentication secret -- will default to no authentication if unspecified.
+    #[structopt(short = "a", long = "auth-secret")]
+    pub auth_secret: Option<String>,
+
+    /// The subcommand -- e.g., ping.
+    #[structopt(subcommand)]
+    pub subcommand: Subcommand,
+}
+
+impl ParsecToolApp {
+    /// Run the requested subcommand.
+    pub fn dispatch_subcommand(&self) -> Result<(), ParsecToolError> {
```

Comment:
Nitpicking here but in a future version, if that is ok, `Subcommand` itself could implement the `ParsecToolSubcommand` trait so that you could call `run` on it directly from `main` without needing this method.